server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1779,7 +1780,8 @@ public static ClusterState stateWithoutSnapshot(ClusterState state, Snapshot sna
                             .entrySet()) {
                             final ShardSnapshotStatus shardState = finishedShardEntry.getValue();
                             if (shardState.state() == ShardState.SUCCESS
-                                && previousEntry.shardsByRepoShardId().containsKey(finishedShardEntry.getKey())) {
+                                && previousEntry.shardsByRepoShardId().containsKey(finishedShardEntry.getKey())
+                                && indexExists(removedEntry, state.metadata(), finishedShardEntry.getKey().indexName())) {
```

Comment:
Hmm yeah I tried going down this path too, there's some potential inconsistencies here still. In fact I'd rather look directly at `org.elasticsearch.repositories.FinalizeSnapshotContext#updatedShardGenerations` when deciding what to do with all the other shard generations. But note that we also call this from `org.elasticsearch.snapshots.SnapshotsService#removeFailedSnapshotFromClusterState` which doesn't have a corresponding `FinalizeSnapshotContext`, and I got stuck trying to work out what `ShardGenerations` to use in that case.

Reply:
Yeah, it would be nice to use the `updatedShardGenerations`. It looks like `removeFailedSnapshotFromClusterState` has 3 usages, where only one happens within finalize. For that one, we could grab the shard generations to use.

The other 2 usages seems to be clone error handling related. AFAICS, we could do no shard assignment updates there, since it happens prior to finalizing? That would IIUC then use the existing shard generations rather than a potential newer one for those previous snapshots, but given that the clone is being failed/removed, that seems just fine?

Reply:
This is a fine observation. Along the line of what Henning suggested, I think we could provide a more consistent `metadata` for the `removeFailedSnapshotFromClusterState` use case as well:
1. For the usage in [SnapshotsService#endSnapshot](https://github.com/elastic/elasticsearch/blob/1d525752af4f4c768d3a13604f13e5d3005d284a/server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java#L1315), we could use the `metadata` argument passed in. This is the same metadata used later for computing updated shard generations (`buildGenerations`).
2. For the usage in [SnapshotsService#handleFinalizationFailure](https://github.com/elastic/elasticsearch/blob/1d525752af4f4c768d3a13604f13e5d3005d284a/server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java#L1584), the method is called only within `finalizeSnapshotEntry` which also has `metadata` from the argument.
3. Lastly, for the usage in [SnapshotsService#startCloning](https://github.com/elastic/elasticsearch/blob/1d525752af4f4c768d3a13604f13e5d3005d284a/server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java#L453), the call seems to happen before the actual snapshot. Hence it should not be able to update any shard assignments. Honestly, I am not 100% sure about this item. Please let me know if my understanding is off.